### PR TITLE
DOC: document array methods

### LINKED ISSUE #14563: Document jax.Array methods & attributes
Currently there is no documentation for jax Array attributes and methods.

#14474 was an attempt at this, but it caused issues and had to be rolled-back in #510467038

We can't document `ArrayImpl` directly, because its import path does not match its `__module__` and `__name__`, and so sphinx treats it as an alias.

I'm not sure the best way to proceed.

### PR DESCRIPTION
**Pull request by jakevdp**
Fixes https://github.com/google/jax/issues/14563

Preview new docs here: https://jax--15012.org.readthedocs.build/en/15012/jax.numpy.html#class-reference

### CODE REVIEW

File: docs/jax.numpy.rst [hunk 1]
```diff
@@ -498,3 +498,13 @@ This is because in general, pickling and unpickling may take place in different
...
 to the device IDs of another. If :mod:`pickle` is used in traced/JIT-compiled code,
 it will result in a :class:`~jax.errors.ConcretizationTypeError`.
+
+Class Reference
+~~~~~~~~~~~~~~~
+
+.. autoclass:: jax.Array
+   :noindex:
+
+.. autoclass:: jaxlib.xla_extension.ArrayImpl
```
[the remark this answers is not in the document]
Comment (jakevdp) [reply docs/jax.numpy.rst:508]: The only way to document `ArrayImpl` methods is to document `ArrayImpl`, so yes this is the core change in this PR 😁

File: jax/_src/array.py [hunk 2]
```diff
@@ -133,6 +133,20 @@ def _is_reduced_on_dim(idx):
 
 
 class ArrayImpl(basearray.Array):
+  """JAX Array Implementation
+
+  Most users should never reference ``ArrayImpl`` directly.
```
Comment (yashk2810) [inline jax/_src/array.py:138]: Make this `No` user?

File: docs/jax.numpy.rst (hunk 1 above)
[the remark this answers is not in the document]
Comment (jakevdp) [reply docs/jax.numpy.rst:508]: An alternative is what I tried in #14474 (making `jax.Array` a true abstract base class with abstract methods defined), and then we could document these methods via `jax.Array`. This has some subtle side-effects regarding how tracers are implemented, so I'm wary of going that route

File: jax/_src/array.py (hunk 2 above)
Comment (jakevdp) [reply jax/_src/array.py]: Done

File: docs/jax.numpy.rst (hunk 1 above)
[the remark this answers is not in the document]
Comment (jakevdp) [reply docs/jax.numpy.rst:508]: In either case, I think it's beneficial to document the `ArrayImpl` type. Whether we document it or not, users are going to find the type by doing `print(type(x))` – I think it's better for the `ArrayImpl` type to appear in the docs with this explicit messaging, than for us to ignore it and hope users infer our intent that it not be used directly.

**Comment by jblespiau** [comment]
Is this still planned to be submitted?

**Comment by jakevdp** [comment]
I think this is no longer necessary because we now document `jax.Array` as an abstract base class: https://jax.readthedocs.io/en/latest/_autosummary/jax.Array.html

**Pull request closed without merging**